libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,584 @@
+    update(position_ok, current_pos);
+}
+
+void AP_SafeRTL::update(bool position_ok, const Vector3f& current_pos)
+{
+    if (!_active) {
+        return;
+    }
+
+    if (position_ok) {
+        _last_good_position_ms = AP_HAL::millis();
+    } else {
+        if (AP_HAL::millis() - _last_good_position_ms > SAFERTL_BAD_POSITION_TIMEOUT) {
+            _active = false;
+            log_action(SRTL_DEACTIVATED_BAD_POSITION);
+            gcs().send_text(MAV_SEVERITY_WARNING,"SafeRTL deactivated: bad position");
+        }
+        return;
+    }
+
+    // it's important to do the cleanup before adding the point, because appending a point will reset the cleanup methods,
+    // so there will not be anything to clean up immediately after adding a point.
+    // The cleanup usually returns immediately. If it decides to actually perform the cleanup, it takes about 100us.
+    if (!routine_cleanup()) {
+        _active = false;
+        log_action(SRTL_DEACTIVATED_CLEANUP_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING,"SafeRTL deactivated: path cleanup failed");
+        return;
+    }
+
+    if (position_ok) {
+        // append the new point, if we have traveled far enough
+        if (HYPOT(current_pos, _path[_last_index]) > _accuracy) {
+            // add the breadcrumb
+            _path[++_last_index] = current_pos;
+
+            log_action(SRTL_POINT_ADD, current_pos);
+
+            // if cleanup algorithms are finished (and therefore not running), reset them
+            if (_simplify_complete) {
+                restart_simplification();
+            }
+            if (_prune_complete) {
+                restart_pruning();
+            }
+        }
+    }
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
+*/
+bool AP_SafeRTL::thorough_cleanup()
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
+    }
+
+    // check if we are ready to perform cleanup.  this should be called just before thorough_cleanup just before initiating the RTL
+    if (!_prune_complete || !_simplify_complete) {
+        return false;
+    }
+
+    // apply simplification
+    zero_points_by_simplify_bitmask();
+
+    // apply pruning, prune every single loop
+    zero_points_by_loops(SAFERTL_POINTS_MAX);
+
+    remove_empty_points();
+
+    // end by resetting the state of the cleanup methods.
+    reset_simplification();
+    reset_pruning();
+
+    return true;
+}
+
+/**
+*   Returns true if the list is empty after popping this point.
+*/
+bool AP_SafeRTL::pop_point(Vector3f& point)
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
```

Comment:
This is problematic in that the error condition is indistinguishable from the normal return value of false. If this is ever hit it the caller will think they have a valid location to go to, but it will just be garbage on the stack. At a minimum if this check is needed then it should set the point to be the current location and the vehicle will at least be stuck there.